File: cpp/src/arrow/json/reader-test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <gtest/gtest.h>

#include "arrow/io/interfaces.h"
#include "arrow/json/options.h"
#include "arrow/json/reader.h"
#include "arrow/json/test-common.h"
#include "arrow/table.h"
#include "arrow/testing/gtest_util.h"

namespace arrow {
namespace json {

using util::string_view;

using internal::checked_cast;

static std::string scalars_only_src() {
  return R"(
    { "hello": 3.5, "world": false, "yo": "thing" }
    { "hello": 3.25, "world": null }
    { "hello": 3.125, "world": null, "yo": "\u5fcd" }
    { "hello": 0.0, "world": true, "yo": null }
  )";
}

static std::string nested_src() {
  return R"(
    { "hello": 3.5, "world": false, "yo": "thing", "arr": [1, 2, 3], "nuf": {} }
    { "hello": 3.25, "world": null, "arr": [2], "nuf": null }
    { "hello": 3.125, "world": null, "yo": "\u5fcd", "arr": [], "nuf": { "ps": 78 } }
    { "hello": 0.0, "world": true, "yo": null, "arr": null, "nuf": { "ps": 90 } }
  )";
}

class ReaderTest : public ::testing::TestWithParam<bool> {
 public:
  void SetUpReader() {
    read_options_.use_threads = GetParam();
    ASSERT_OK(TableReader::Make(default_memory_pool(), input_, read_options_,
                                parse_options_, &reader_));
  }

  void SetUpReader(util::string_view input) {
    ASSERT_OK(MakeStream(input, &input_));
    SetUpReader();
  }

  std::shared_ptr<Column> ColumnFromJSON(const std::shared_ptr<Field>& field,
                                         const std::string& data) {
    return std::make_shared<Column>(field, ArrayFromJSON(field->type(), data));
  }

  std::shared_ptr<Column> ColumnFromJSON(const std::shared_ptr<Field>& field,
                                         const std::vector<std::string>& data) {
    ArrayVector chunks(data.size());
    for (size_t i = 0; i < chunks.size(); ++i) {
      chunks[i] = ArrayFromJSON(field->type(), data[i]);
    }
    return std::make_shared<Column>(field, std::move(chunks));
  }

  ParseOptions parse_options_ = ParseOptions::Defaults();
  ReadOptions read_options_ = ReadOptions::Defaults();
  std::shared_ptr<io::InputStream> input_;
  std::shared_ptr<TableReader> reader_;
  std::shared_ptr<Table> table_;
};

INSTANTIATE_TEST_CASE_P(ReaderTest, ReaderTest, ::testing::Values(false, true));

TEST_P(ReaderTest, Empty) {
  SetUpReader("{}\n{}\n");
  ASSERT_OK(reader_->Read(&table_));

  auto expected_table = Table::Make(schema({}), ArrayVector(), 2);
  AssertTablesEqual(*expected_table, *table_);
}

TEST_P(ReaderTest, Basics) {
  parse_options_.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;
  auto src = scalars_only_src();
  SetUpReader(src);
  ASSERT_OK(reader_->Read(&table_));

  auto expected_table = Table::Make({
      ColumnFromJSON(field("hello", float64()), "[3.5, 3.25, 3.125, 0.0]"),
      ColumnFromJSON(field("world", boolean()), "[false, null, null, true]"),
      ColumnFromJSON(field("yo", utf8()), "[\"thing\", null, \"\xe5\xbf\x8d\", null]"),
  });
  AssertTablesEqual(*expected_table, *table_);
}

TEST_P(ReaderTest, Nested) {
  parse_options_.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;
  auto src = nested_src();
  SetUpReader(src);
  ASSERT_OK(reader_->Read(&table_));

  auto expected_table = Table::Make({
      ColumnFromJSON(field("hello", float64()), "[3.5, 3.25, 3.125, 0.0]"),
      ColumnFromJSON(field("world", boolean()), "[false, null, null, true]"),
      ColumnFromJSON(field("yo", utf8()), "[\"thing\", null, \"\xe5\xbf\x8d\", null]"),
      ColumnFromJSON(field("arr", list(int64())), R"([[1, 2, 3], [2], [], null])"),
      ColumnFromJSON(field("nuf", struct_({field("ps", int64())})),
                     R"([{"ps":null}, null, {"ps":78}, {"ps":90}])"),
  });
  AssertTablesEqual(*expected_table, *table_);
}

TEST_P(ReaderTest, PartialSchema) {
  parse_options_.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;
  parse_options_.explicit_schema =
      schema({field("nuf", struct_({field("absent", date32())})),
              field("arr", list(float32()))});
  auto src = nested_src();
  SetUpReader(src);
  ASSERT_OK(reader_->Read(&table_));

  auto expected_table = Table::Make({
      // NB: explicitly declared fields will appear first
      ColumnFromJSON(
          field("nuf", struct_({field("absent", date32()), field("ps", int64())})),
          R"([{"absent":null,"ps":null}, null, {"absent":null,"ps":78}, {"absent":null,"ps":90}])"),
      ColumnFromJSON(field("arr", list(float32())), R"([[1, 2, 3], [2], [], null])"),
      // ...followed by undeclared fields
      ColumnFromJSON(field("hello", float64()), "[3.5, 3.25, 3.125, 0.0]"),
      ColumnFromJSON(field("world", boolean()), "[false, null, null, true]"),
      ColumnFromJSON(field("yo", utf8()), "[\"thing\", null, \"\xe5\xbf\x8d\", null]"),
  });
  AssertTablesEqual(*expected_table, *table_);
}

TEST_P(ReaderTest, TypeInference) {
  parse_options_.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;
  SetUpReader(R"(
    {"ts":null, "f": null}
    {"ts":"1970-01-01", "f": 3}
    {"ts":"2018-11-13 17:11:10", "f":3.125}
    )");
  ASSERT_OK(reader_->Read(&table_));

  auto expected_table =
      Table::Make({ColumnFromJSON(field("ts", timestamp(TimeUnit::SECOND)),
                                  R"([null, "1970-01-01", "2018-11-13 17:11:10"])"),
                   ColumnFromJSON(field("f", float64()), R"([null, 3, 3.125])")});
  AssertTablesEqual(*expected_table, *table_);
}

TEST_P(ReaderTest, MutlipleChunks) {
  parse_options_.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;

  auto src = scalars_only_src();
  read_options_.block_size = static_cast<int>(src.length() / 3);

  SetUpReader(src);
  ASSERT_OK(reader_->Read(&table_));

  // there is an empty chunk because the last block of the file is "  "
  auto expected_table = Table::Make({
      ColumnFromJSON(field("hello", float64()),
                     {"[3.5]", "[3.25]", "[3.125, 0.0]", "[]"}),
      ColumnFromJSON(field("world", boolean()),
                     {"[false]", "[null]", "[null, true]", "[]"}),
      ColumnFromJSON(field("yo", utf8()),
                     {"[\"thing\"]", "[null]", "[\"\xe5\xbf\x8d\", null]", "[]"}),
  });
  AssertTablesEqual(*expected_table, *table_);
}

template <typename T>
std::string RowsOfOneColumn(string_view name, std::initializer_list<T> values,
                            decltype(std::to_string(*values.begin()))* = nullptr) {
  std::stringstream ss;
  for (auto value : values) {
    ss << R"({")" << name << R"(":)" << std::to_string(value) << "}\n";
  }
  return ss.str();
}

TEST(ReaderTest, MultipleChunksParallel) {
  int64_t count = 1 << 20;

  ParseOptions parse_options;
  parse_options.unexpected_field_behavior = UnexpectedFieldBehavior::InferType;
  ReadOptions read_options;
  read_options.block_size =
      static_cast<int>(count / 2);  // there will be about two dozen blocks

  std::string json;
  for (int i = 0; i < count; ++i) {
    json += "{\"a\":" + std::to_string(i) + "}\n";
  }
  std::shared_ptr<io::InputStream> input;
  std::shared_ptr<Table> serial, threaded;
  std::shared_ptr<TableReader> reader;

  read_options.use_threads = true;
  ASSERT_OK(MakeStream(json, &input));
  ASSERT_OK(TableReader::Make(default_memory_pool(), input, read_options, parse_options,
                              &reader));
  ASSERT_OK(reader->Read(&threaded));

  read_options.use_threads = false;
  ASSERT_OK(MakeStream(json, &input));
  ASSERT_OK(TableReader::Make(default_memory_pool(), input, read_options, parse_options,
                              &reader));
  ASSERT_OK(reader->Read(&serial));

  ASSERT_EQ(serial->column(0)->type()->id(), Type::INT64);
  int expected = 0;
  for (auto chunk : serial->column(0)->data()->chunks()) {
    for (int64_t i = 0; i < chunk->length(); ++i) {
      ASSERT_EQ(checked_cast<const Int64Array*>(chunk.get())->GetView(i), expected)
          << " at index " << i;
      ++expected;
    }
  }

  // std::cout << serial->column(0)->data()->num_chunks() << " chunks, " << json.size()
  //           << " bytes" << std::endl;

  AssertTablesEqual(*serial, *threaded);
}

}  // namespace json
}  // namespace arrow

File: go/arrow/internal/arrjson/reader_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package arrjson // import "github.com/apache/arrow/go/arrow/internal/arrjson"

import (
	"io"
	"log"
	"strings"
	"testing"

	"github.com/apache/arrow/go/arrow"
	"github.com/apache/arrow/go/arrow/array"
	"github.com/apache/arrow/go/arrow/memory"
)

func TestReader(t *testing.T) {
	for _, tc := range []struct {
		name   string
		json   string
		schema *arrow.Schema
		want   []array.Record
	}{
		{
			name: "example1",
			json: example1,
			schema: arrow.NewSchema([]arrow.Field{
				{Name: "foo", Type: arrow.PrimitiveTypes.Int32, Nullable: true},
			}, nil),
			want: nil,
		},
		{
			name: "example2",
			json: example2,
			schema: arrow.NewSchema([]arrow.Field{
				{Name: "foo", Type: arrow.PrimitiveTypes.Int32, Nullable: true},
				{Name: "bar", Type: arrow.PrimitiveTypes.Float64, Nullable: true},
			}, nil),
			want: nil,
		},
		{
			name: "example3",
			json: example3,
			schema: arrow.NewSchema([]arrow.Field{
				{Name: "list_nullable", Type: arrow.ListOf(arrow.PrimitiveTypes.Int32), Nullable: true},
				{Name: "struct_nullable", Type: arrow.StructOf([]arrow.Field{
					{Name: "f1", Type: arrow.PrimitiveTypes.Int32, Nullable: true},
					{Name: "f2", Type: arrow.BinaryTypes.String, Nullable: true},
				}...), Nullable: true},
			}, nil),
			want: nil,
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
			defer mem.AssertSize(t, 0)

			r, err := NewReader(strings.NewReader(tc.json), WithAllocator(mem))
			if err != nil {
				t.Fatal(err)
			}
			defer r.Release()

			if got, want := r.Schema(), tc.schema; !got.Equal(want) {
				t.Fatalf("invalid schema\ngot:\n%v\nwant:\n%v\n", got, want)
			}
			for {
				rec, err := r.Read()
				if err == io.EOF {
					break
				}
				if err != nil {
					t.Fatal(err)
				}
				log.Printf("record:")
				for j, col := range rec.Columns() {
					log.Printf("col[%d][%s]: %v\n", j, rec.Schema().Field(j).Name, col)
				}
			}
		})
	}
}

const (
	example1 = `{
  "schema": {
    "fields": [
      {
        "name": "foo",
        "type": {"name": "int", "isSigned": true, "bitWidth": 32},
        "nullable": true, "children": [],
        "typeLayout": {
          "vectors": [
            {"type": "VALIDITY", "typeBitWidth": 1},
            {"type": "DATA", "typeBitWidth": 32}
          ]
        }
      }
    ]
  },
  "batches": [
    {
      "count": 5,
      "columns": [
        {
          "name": "foo",
          "count": 5,
          "DATA": [1, 2, 3, 4, 5],
          "VALIDITY": [1, 0, 1, 1, 1]
        }
      ]
    }
  ]
}`
	example2 = `{
  "schema": {
    "fields": [
      {
        "name": "foo",
        "type": {"name": "int", "isSigned": true, "bitWidth": 32},
        "nullable": true, "children": [],
        "typeLayout": {
          "vectors": [
            {"type": "VALIDITY", "typeBitWidth": 1},
            {"type": "DATA", "typeBitWidth": 32}
          ]
        }
      },
      {
        "name": "bar",
        "type": {"name": "floatingpoint", "precision": "DOUBLE"},
        "nullable": true, "children": [],
        "typeLayout": {
          "vectors": [
            {"type": "VALIDITY", "typeBitWidth": 1},
            {"type": "DATA", "typeBitWidth": 64}
          ]
        }
      }
    ]
  },
  "batches": [
    {
      "count": 5,
      "columns": [
        {
          "name": "foo",
          "count": 5,
          "DATA": [1, 2, 3, 4, 5],
          "VALIDITY": [1, 0, 1, 1, 1]
        },
        {
          "name": "bar",
          "count": 5,
          "DATA": [1.0, 2.0, 3.0, 4.0, 5.0],
          "VALIDITY": [1, 0, 0, 1, 1]
        }
      ]
    },
    {
      "count": 4,
      "columns": [
        {
          "name": "foo",
          "count": 4,
          "DATA": [1, 2, 3, 4],
          "VALIDITY": [1, 0, 1, 1]
        },
        {
          "name": "bar",
          "count": 4,
          "DATA": [1.0, 2.0, 3.0, 4.0],
          "VALIDITY": [1, 0, 0, 1]
        }
      ]
    }
  ]
}`

	example3 = `{
  "schema": {
    "fields": [
      {
        "name": "list_nullable",
        "type": {
          "name": "list"
        },
        "nullable": true,
        "children": [
          {
            "name": "item",
            "type": {
              "name": "int",
              "isSigned": true,
              "bitWidth": 32
            },
            "nullable": true,
            "children": []
          }
        ]
      },
      {
        "name": "struct_nullable",
        "type": {
          "name": "struct"
        },
        "nullable": true,
        "children": [
          {
            "name": "f1",
            "type": {
              "name": "int",
              "isSigned": true,
              "bitWidth": 32
            },
            "nullable": true,
            "children": []
          },
          {
            "name": "f2",
            "type": {
              "name": "utf8"
            },
            "nullable": true,
            "children": []
          }
        ]
      }
    ]
  },
  "batches": [
    {
      "count": 7,
      "columns": [
        {
          "name": "list_nullable",
          "count": 7,
          "VALIDITY": [
            1,
            1,
            0,
            1,
            0,
            0,
            0
          ],
          "OFFSET": [
            0,
            4,
            8,
            8,
            12,
            12,
            12,
            12
          ],
          "children": [
            {
              "name": "item",
              "count": 12,
              "VALIDITY": [
                0,
                1,
                1,
                1,
                1,
                1,
                0,
                0,
                1,
                1,
                1,
                1
              ],
              "DATA": [
                -1801578480,
                -340485928,
                -259730121,
                -1430069283,
                900706715,
                -630883814,
                -1732791456,
                237606,
                -1410464938,
                -854246234,
                -818241763,
                893338527
              ]
            }
          ]
        },
        {
          "name": "struct_nullable",
          "count": 7,
          "VALIDITY": [
            0,
            0,
            1,
            1,
            1,
            1,
            0
          ],
          "children": [
            {
              "name": "f1",
              "count": 7,
              "VALIDITY": [
                0,
                0,
                0,
                1,
                0,
                0,
                1
              ],
              "DATA": [
                -946762941,
                970770270,
                836792346,
                -724187464,
                1289964601,
                -310139442,
                -1169064100
              ]
            },
            {
              "name": "f2",
              "count": 7,
              "VALIDITY": [
                1,
                1,
                1,
                0,
                0,
                1,
                1
              ],
              "OFFSET": [
                0,
                7,
                14,
                21,
                21,
                21,
                28,
                35
              ],
              "DATA": [
                "85r6rN5",
                "jrDJR09",
                "YfZLF07",
                "",
                "",
                "8jDK2tj",
                "146gCTu"
              ]
            }
          ]
        }
      ]
    },
    {
      "count": 10,
      "columns": [
        {
          "name": "list_nullable",
          "count": 10,
          "VALIDITY": [
            0,
            1,
            1,
            0,
            1,
            1,
            1,
            0,
            0,
            1
          ],
          "OFFSET": [
            0,
            0,
            4,
            8,
            8,
            9,
            11,
            11,
            11,
            11,
            11
          ],
          "children": [
            {
              "name": "item",
              "count": 11,
              "VALIDITY": [
                0,
                0,
                1,
                0,
                0,
                1,
                0,
                1,
                1,
                0,
                0
              ],
              "DATA": [
                103442105,
                2117865335,
                2968015,
                -939072509,
                1200505245,
                1556114914,
                -1326977325,
                139341080,
                -1938030331,
                1500012688,
                955080825
              ]
            }
          ]
        },
        {
          "name": "struct_nullable",
          "count": 10,
          "VALIDITY": [
            0,
            0,
            0,
            1,
            0,
            1,
            0,
            1,
            0,
            1
          ],
          "children": [
            {
              "name": "f1",
              "count": 10,
              "VALIDITY": [
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                0,
                0,
                1
              ],
              "DATA": [
                1689993413,
                -379972520,
                441054176,
                -906001202,
                -968548682,
                -450386762,
                916507802,
                -1408175154,
                152570373,
                2072407432
              ]
            },
            {
              "name": "f2",
              "count": 10,
              "VALIDITY": [
                0,
                0,
                0,
                1,
                1,
                1,
                1,
                0,
                0,
                1
              ],
              "OFFSET": [
                0,
                0,
                0,
                0,
                7,
                14,
                21,
                28,
                28,
                28,
                35
              ],
              "DATA": [
                "",
                "",
                "",
                "DhKdnsR",
                "20Zs1ez",
                "ky5eL4K",
                "UkTkWDX",
                "",
                "",
                "OR0xSBn"
              ]
            }
          ]
        }
      ]
    }
  ]
}`
)
